ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -0,0 +1,350 @@
+import com.github.ambry.utils.Utils;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import org.apache.commons.lang3.tuple.Triple;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Compression service is thread-safe singleton that apply compression and decompression.
+ * Its implementation is backed by the CompressionMap in the common library.
+ */
+public class CompressionService {
+
+  private static final Logger logger = LoggerFactory.getLogger(CompressionService.class);
+
+  // MB/sec = (byte/microsecond * MB/1024*1024bytes * 1000000Microseconds/sec) = 1 MB/1.048576 sec
+  private static final double BytePerMicrosecondToMBPerSec = 1/1.048576;
+
+  private final CompressionConfig compressionConfig;
+  private final CompressionMetrics compressionMetrics;
+
+  // When a new compression algorithm is introduced, add the new algorithm here.
+  // WARNING: Do not remove compression algorithms from this list.  See {@code Compression} interface for detail.
+  private final CompressionMap allCompressions = CompressionMap.of(new ZstdCompression(), new LZ4Compression());
+
+  /**
+   * Create an instance of the compression service using the compression config and metrics.
+   * The default compressor is loaded from config, if available.  If not available, ZStd is the default.
+   * @param config Compression config.
+   * @param metrics Compression metrics.
+   */
+  public CompressionService(CompressionConfig config, CompressionMetrics metrics) {
+    compressionConfig = config;
+    compressionMetrics = metrics;
+  }
+
+  /**
+   * Get the compression map that contains all compression algorithms.
+   * @return The compression map.
+   */
+  public CompressionMap getAllCompressions() {
+    return allCompressions;
+  }
+
+  /**
+   * Check whether the blob is compressible based on the blob properties.
+   *
+   * @param blobProperties The blob property that is common to all chunks. It helps to detect whether blob is
+   *                       compressible or not.
+   * @return True if blob is compressible based on the blob properties; FALSE if blob is not compressible.
+   */
+  public boolean isBlobCompressible(BlobProperties blobProperties) {
+    Objects.requireNonNull(blobProperties, "blobProperties cannot be null.");
+
+    // Return false if compression is disabled.  No need to emit metrics.
+    if (!compressionConfig.isCompressionEnabled) {
+      return false;
+    }
+
+    // Check content-encoding. BlobProperty.ContentEncoding contains a string, do not apply compression.
+    // Note: Calling getBlobProperties() throws if blob properties is null, so it never returns null.
+    String contentEncoding = blobProperties.getContentEncoding();
+    if (compressionConfig.isSkipWithContentEncoding && !Utils.isNullOrEmpty(contentEncoding)) {
+      logger.info("No compression applied because compression content is encoded with " + contentEncoding);
```

Comment:
This will overwhelm the log

Reply:
Let's talk about how we use the logs.  It's subjective how much to log.  When things go well, it seems to log too much.  When something goes wrong, we wish we have more logs.  Would it help if I change this to trace?

Reply:
Yup trace is better. If something goes wrong, you can enable tracing in prod.